Prevent the `AnnotationStorage`-data from being "silently" modified when using the the `getRawValue`-method (PR 14869 follow-up)

Snuffleupagus opened the pull request:
*I missed this during review of PR #14869, sorry about that!*

The way that the `AnnotationStorage.getRawValue`-method is implemented, and then used in the annotation-layer, unfortunately means that it's possible to *modify* the stored value without that correctly triggering the existing "modified"-handling.

This matters since the `lastModified`-getter is being used to compute the cache-key that we use for OperatorList-caching in the API, and without these changes it's thus *possible* that we fail to invalidate the cache as intended.
Given how the `AnnotationStorage.getRawValue`-method is currently being used that's not likely to cause problems, but with any future changes the old code *could* have caused weird/intermittent rendering bugs.

Snuffleupagus committed:
Prevent the `AnnotationStorage`-data from being "silently" modified when using the the `getRawValue`-method (PR 14869 follow-up)

*I missed this during review of PR 14869, sorry about that!*

The way that the `AnnotationStorage.getRawValue`-method is implemented, and then used in the annotation-layer, unfortunately means that it's possible to *modify* the stored value without that correctly triggering the existing "modified"-handling.

This matters since the `lastModified`-getter is being used to compute the cache-key that we use for OperatorList-caching in the API, and without these changes it's thus *possible* that we fail to invalidate the cache as intended.
Given how the `AnnotationStorage.getRawValue`-method is currently being used that's not likely to cause problems, but with any future changes the old code *could* have caused weird/intermittent rendering bugs.

calixteman commented:
It was deliberate (the first name I gave to this method was `getMutableValue`).
The only use of this value is:
https://github.com/mozilla/pdf.js/blob/8135d7ccf6c8789aad2c381d126e276b675ffbe2/src/display/annotation_layer.js#L399

We're in the case where an annotation got some properties thanks to JS but this annotation is not displayed.
So in this case the raw event is stored:
https://github.com/mozilla/pdf.js/blob/8135d7ccf6c8789aad2c381d126e276b675ffbe2/web/pdf_scripting_manager.js#L359

So my idea was to clean-up the data we've in the annotation storage for this annotation.

Snuffleupagus committed:
Update `objectFromMap` to support copying of sub-objects

Similar to the previous patch, this functionality is necessary for the `AnnotationStorage.getAll`-method.

Snuffleupagus commented:
But the value should still be removed, just as before, but in a way that won't cause problems later on.

Snuffleupagus commented:
> It was deliberate (the first name I gave to this method was `getMutableValue`).

Unfortunately, whatever we name it, there's always a risk that someone (even you or me) in the future *accidentally* starts using the method somewhere where it actually matters that'll short-circuit the "modified"-handling.
Hence I think that we need to, somehow, figure out a way to handle this since the kind of rendering bug that the current method *could potentially* lead to may not be obvious or easy to debug.

calixteman commented:
The action itself:
https://github.com/mozilla/pdf.js/blob/8135d7ccf6c8789aad2c381d126e276b675ffbe2/src/display/annotation_layer.js#L397
should have updated if necessary the storage, for example:
https://github.com/mozilla/pdf.js/blob/8135d7ccf6c8789aad2c381d126e276b675ffbe2/src/display/annotation_layer.js#L312

but in the case of a color:
https://github.com/mozilla/pdf.js/blob/8135d7ccf6c8789aad2c381d126e276b675ffbe2/src/display/annotation_layer.js#L352

I think we don't have to update the lastModified property in this case.

But what to do in case there is an action on pageOpen to set a field to the current date: we'll modify the lastModified but there was no action from the user.
In general maybe we should change the lastModified only on a direct user interaction, wdyt ?

Snuffleupagus commented:
> In general maybe we should change the lastModified only on a direct user interaction, wdyt ?

The problem is to *ensure* that this is always safe, since it's much better to wrongly invalidate a cache than keeping stale data alive. (The first case reduces performance, the second one potentially causes rendering bugs.)
Besides, is it even guaranteed in general that *only* user-interation should change `lastModified`?

Snuffleupagus commented:
Another, possibly overall much better method, might be to replace the `lastModified`-method with code that computes a hash of the actual AnnotationStorage-contents since that should prevent any future problems here.

(Note that the current implementation was chosen *only* for its simplicity.)

calixteman commented:
We could here:
https://github.com/mozilla/pdf.js/blob/8135d7ccf6c8789aad2c381d126e276b675ffbe2/web/pdf_scripting_manager.js#L359
pass something like `{ _details: details }`.
and compute a hash based on keys without an underscore as first char.
